File: apps/api/src/solver2/python/time_grid.py
"""Time grid utilities for Solver V2.

Creates a discrete slot grid spanning the store day plus all crew shift bounds
so later modeling stages can reason about contiguous slots, hour buckets, and
per-crew eligibility without duplicating indexing logic.
"""
from __future__ import annotations

from dataclasses import dataclass
from math import ceil
from typing import Any, Dict, List, Optional


@dataclass
class SlotDescriptor:
    index: int
    start_minute: int
    end_minute: int

    @property
    def hour(self) -> int:
        return self.start_minute // 60


@dataclass
class CrewShiftWindow:
    crew_id: str
    start_minute: int
    end_minute: int
    slot_start: int
    slot_end: int  # exclusive

    @property
    def duration_minutes(self) -> int:
        return max(0, self.end_minute - self.start_minute)

    @property
    def slot_indexes(self) -> range:
        return range(self.slot_start, self.slot_end)


@dataclass
class TimeGrid:
    base_slot_minutes: int
    open_minutes: int
    close_minutes: int
    day_start_minute: int
    day_end_minute: int
    slots: List[SlotDescriptor]
    crew_windows: Dict[str, CrewShiftWindow]

    @property
    def slots_per_day(self) -> int:
        return len(self.slots)

    def summary(self) -> Dict[str, Any]:
        return {
            "base_slot_minutes": self.base_slot_minutes,
            "slots_per_day": self.slots_per_day,
            "crew_windows": len(self.crew_windows),
        }

    def minute_to_slot_index(self, minute: int) -> Optional[int]:
        if minute < self.day_start_minute or minute >= self.day_end_minute:
            return None
        offset = minute - self.day_start_minute
        return offset // self.base_slot_minutes

    def slot_index_to_minute(self, slot_index: int) -> int:
        return self.day_start_minute + slot_index * self.base_slot_minutes

    def block_bounds(self, slot_index: int, block_size_slots: int) -> Optional[tuple[int, int]]:
        if slot_index < 0 or slot_index >= len(self.slots):
            return None
        end_index = slot_index + block_size_slots - 1
        if end_index >= len(self.slots):
            return None
        return (
            self.slots[slot_index].start_minute,
            self.slots[end_index].end_minute,
        )

    def is_within_store_hours(self, start_minute: int, end_minute: int) -> bool:
        return start_minute >= self.open_minutes and end_minute <= self.close_minutes


def build_time_grid(solver_input: Dict[str, Any]) -> TimeGrid:
    store = solver_input.get("store", {})
    crew_records = solver_input.get("crew", [])
    base_slot_minutes = int(store.get("baseSlotMinutes", 30))
    open_minutes = int(store.get("openMinutesFromMidnight", 0))
    close_minutes = int(store.get("closeMinutesFromMidnight", 24 * 60))

    if crew_records:
        earliest_shift = min(int(crew.get("shiftStartMin", open_minutes)) for crew in crew_records)
        latest_shift = max(int(crew.get("shiftEndMin", close_minutes)) for crew in crew_records)
    else:
        earliest_shift = open_minutes
        latest_shift = close_minutes

    day_start_minute = min(open_minutes, earliest_shift)
    day_end_minute = max(close_minutes, latest_shift)
    if day_end_minute <= day_start_minute:
        day_end_minute = day_start_minute + base_slot_minutes

    slot_count = ceil((day_end_minute - day_start_minute) / base_slot_minutes)
    slots: List[SlotDescriptor] = []
    for index in range(slot_count):
        start = day_start_minute + index * base_slot_minutes
        end = start + base_slot_minutes
        slots.append(SlotDescriptor(index=index, start_minute=start, end_minute=end))

    crew_windows: Dict[str, CrewShiftWindow] = {}
    for crew in crew_records:
        start_minute = int(crew.get("shiftStartMin", day_start_minute))
        end_minute = int(crew.get("shiftEndMin", start_minute))
        if end_minute <= start_minute:
            continue
        slot_start = max(0, (start_minute - day_start_minute) // base_slot_minutes)
        slot_end = min(slot_count, ceil((end_minute - day_start_minute) / base_slot_minutes))
        if slot_end <= slot_start:
            continue
        crew_windows[crew["id"]] = CrewShiftWindow(
            crew_id=crew["id"],
            start_minute=start_minute,
            end_minute=end_minute,
            slot_start=slot_start,
            slot_end=slot_end,
        )

    return TimeGrid(
        base_slot_minutes=base_slot_minutes,
        open_minutes=open_minutes,
        close_minutes=close_minutes,
        day_start_minute=day_start_minute,
        day_end_minute=day_end_minute,
        slots=slots,
        crew_windows=crew_windows,
    )

File: apps/api/src/solver2/time-grid.ts
// Shared timeline helpers that normalize store + crew minutes into slot indexes the
// rest of solver2 can reason about deterministically.
import type { CrewDescriptor, StoreDescriptor } from './types';

// A normalized view of time expressed in equal slots that wrap the entire
// scheduling horizon (store hours ± crew shift padding).
export interface TimeGrid {
  slotMinutes: number;
  startMinute: number;
  endMinute: number;
  slotCount: number;
  storeOpenMinute: number;
  storeCloseMinute: number;
  storeOpenSlotIndex: number;
  storeCloseSlotIndex: number;
}

// Basic math helpers used throughout the module.
const clamp = (value: number, min: number, max: number) => Math.min(Math.max(value, min), max);
const alignDown = (value: number, step: number) => Math.floor(value / step) * step;
const alignUp = (value: number, step: number) => Math.ceil(value / step) * step;

// Build the canonical slot grid by stretching from the earliest crew start to the
// latest crew end, padding outward to whole-slot boundaries so every other helper
// can safely snap minutes ↔︎ slot indexes.
export function buildTimeGrid(store: StoreDescriptor, crew: CrewDescriptor[]): TimeGrid {
  if (store.baseSlotMinutes <= 0) {
    throw new Error('Store.baseSlotMinutes must be positive');
  }

  const slotMinutes = store.baseSlotMinutes;
  const earliestCrewStart = crew.length
    ? crew.reduce((min, c) => Math.min(min, c.shiftStartMin), Number.POSITIVE_INFINITY)
    : store.openMinutesFromMidnight;
  const latestCrewEnd = crew.length
    ? crew.reduce((max, c) => Math.max(max, c.shiftEndMin), Number.NEGATIVE_INFINITY)
    : store.closeMinutesFromMidnight;

  const gridStart = alignDown(Math.min(store.openMinutesFromMidnight, earliestCrewStart), slotMinutes);
  const gridEnd = alignUp(Math.max(store.closeMinutesFromMidnight, latestCrewEnd), slotMinutes);

  if (!Number.isFinite(gridStart) || !Number.isFinite(gridEnd) || gridEnd <= gridStart) {
    throw new Error('Invalid time bounds for store/crew data');
  }

  const slotCount = Math.round((gridEnd - gridStart) / slotMinutes);

  const grid: TimeGrid = {
    slotMinutes,
    startMinute: gridStart,
    endMinute: gridEnd,
    slotCount,
    storeOpenMinute: store.openMinutesFromMidnight,
    storeCloseMinute: store.closeMinutesFromMidnight,
    storeOpenSlotIndex: 0,
    storeCloseSlotIndex: 0,
  };

  grid.storeOpenSlotIndex = minutesToSlotIndex(store.openMinutesFromMidnight, grid, 'floor');
  grid.storeCloseSlotIndex = minutesToSlotIndex(store.closeMinutesFromMidnight, grid, 'ceil');

  return grid;
}

export function minutesToSlotIndex(
  minute: number,
  grid: TimeGrid,
  rounding: 'floor' | 'ceil' = 'floor'
): number {
  // Translate a minute offset into the nearest slot index using the requested
  // rounding strategy, then clamp the result so callers never see negative
  // indexes or values past the grid end.
  const relative = (minute - grid.startMinute) / grid.slotMinutes;
  const rounded = rounding === 'floor' ? Math.floor(relative) : Math.ceil(relative);
  return clamp(rounded, 0, grid.slotCount);
}

export function slotIndexToMinute(slotIndex: number, grid: TimeGrid): number {
  // Inverse of minutesToSlotIndex – returns the minute value at the start of a
  // slot, again clamping indexes to stay within bounds.
  const clampedIndex = clamp(slotIndex, 0, grid.slotCount);
  return grid.startMinute + clampedIndex * grid.slotMinutes;
}

export function enumerateSlotsBetween(
  startMinute: number,
  endMinute: number,
  grid: TimeGrid
): number[] {
  // Generate the discrete slot indexes covering a minute interval. Useful for
  // iterating coverage windows or shift segments without re-deriving math.
  if (endMinute <= startMinute) {
    return [];
  }
  const startSlot = minutesToSlotIndex(startMinute, grid, 'floor');
  const endSlot = minutesToSlotIndex(endMinute, grid, 'ceil');
  const slots: number[] = [];
  for (let slot = startSlot; slot < endSlot; slot++) {
    slots.push(slot);
  }
  return slots;
}
